File: LibraryBackEnd/Controllers/StatisticsController.cs
using LibraryBackEnd.Core.Services.Interface;
using System.Web.Http;

namespace LibraryBackEnd.Controllers
{
    [Authorize]
    [RoutePrefix("api/statistics")]
    public class StatisticsController : ApiController
    {
        private IBookService _bookService;
        private IStudentService _studentService;
        private IPeriodicMagazineService _periodicMagazineService;

        public StatisticsController(IBookService bookService, IStudentService studentService,
            IPeriodicMagazineService periodicMagazineService)
        {
            _bookService = bookService;
            _studentService = studentService;
            _periodicMagazineService = periodicMagazineService;
        }

        [HttpGet]
        [Route("booksInCourse")]
        public IHttpActionResult GetBooksInCourse()
        {
            var courses = _bookService.GetBooksInCourse();
            return Ok(courses);
        }

        [HttpGet]
        [Route("booksByCourse")]
        public IHttpActionResult GetBooksByCourse(string courseName)
        {
            var courses = _bookService.GetBooksByCourse(courseName);
            return Ok(courses);
        }


        [HttpGet]
        [Route("booksInSubject")]
        public IHttpActionResult GetBooksInSubject()
        {
            var subjects = _bookService.GetBooksInSubject();
            return Ok(subjects);
        }

        [HttpGet]
        [Route("booksBySubject")]
        public IHttpActionResult GetBooksBySubject(string subject)
        {
            if (subject.Contains("  "))
                subject = subject.Replace("  ", "++");
            var books = _bookService.GetBooksBySubject(subject);
            return Ok(books);
        }

        [HttpGet]
        [Route("studentsInYear")]
        public IHttpActionResult GetStudentsInYear()
        {
            var students = _studentService.GetStudentsInYear();
            return Ok(students);
        }


        [HttpGet]
        [Route("studentsByYear/{year}")]
        public IHttpActionResult GetStudentsByYear(int year)
        {
            var students = _studentService.GetStudentsByYear(year);
            return Ok(students);
        }

        [HttpGet]
        [Route("studentsInCourse")]
        public IHttpActionResult GetStudentsInCourse()
        {
            var students = _studentService.GetStudentsInCourse();
            return Ok(students);
        }

        [HttpGet]
        [Route("studentsByCourse")]
        public IHttpActionResult GetStudentsByCourse(string courseName)
        {
            var students = _studentService.GetStudentsByCourse(courseName);
            return Ok(students);
        }


        [HttpGet]
        [Route("bookBaughtInYear")]
        public IHttpActionResult GetBookBaughtInYear()
        {
            var books = _bookService.BookBaughtInYear();
            return Ok(books);
        }

        [HttpGet]
        [Route("booksByYear/{year}")]
        public IHttpActionResult GetBooksByYear(int year)
        {
            var books = _bookService.GetBooksByYear(year);
            return Ok(books);
        }

        [HttpGet]
        [Route("booksTitles")]
        public IHttpActionResult GetBooksTitles()
        {
            var titles = _bookService.GetBooksTitles();
            return Ok(titles);
        }

        [HttpGet]
        [Route("booksByTitles")]
        public IHttpActionResult GetBooksByTitle(string title)
        {
            var books = _bookService.GetBooksByTitle(title);
            return Ok(books);
        }

        [HttpGet]
        [Route("bookTypes")]
        public IHttpActionResult GetBookType()
        {
            var bookTypes = _bookService.GetBookTypes();
            return Ok(bookTypes);
        }

        [HttpGet]
        [Route("booksByType")]
        public IHttpActionResult GetBooksByType(string type)
        {
            var books = _bookService.GetBooksByType(type);
            return Ok(books);
        }

        [HttpGet]
        [Route("magazineTitles")]
        public IHttpActionResult GetMagazineTitles()
        {
            var magazines = _periodicMagazineService.GetMagazineTitles();
            return Ok(magazines);
        }

        [HttpGet]
        [Route("magazineByTitle")]
        public IHttpActionResult GetMagazineByTitle(string title)
        {
            if (title.Contains("  "))
                title = title.Replace("  ", "++");

            var magazines = _periodicMagazineService.GetMagazineByTitle(title);
            return Ok(magazines);
        }

        [HttpGet]
        [Route("magazineCourses")]
        public IHttpActionResult GetMagazineCourse()
        {
            var courses = _periodicMagazineService.GetMagazineCourse();
            return Ok(courses);
        }

        [HttpGet]
        [Route("magazineByCourse")]
        public IHttpActionResult GetMagazineByCourse(string course)
        {
            var magazines = _periodicMagazineService.GetMagazineByCourse(course);
            return Ok(magazines);
        }

        [HttpGet]
        [Route("magazineTypes")]
        public IHttpActionResult GetMagazineTypes()
        {
            var types = _periodicMagazineService.GetMagazineTypes();
            return Ok(types);
        }

        [HttpGet]
        [Route("magazineByType")]
        public IHttpActionResult GetMagazineByTypes(string type)
        {

            var magazines = _periodicMagazineService.GetMagazineByTypes(type);
            return Ok(magazines);
        }

        [HttpGet]
        [Route("magazinePeriodicity")]
        public IHttpActionResult GetMagazinePeriodicity()
        {

            var periodicity = _periodicMagazineService.GetMagazinePeriodicity();
            return Ok(periodicity);
        }

        [HttpGet]
        [Route("magazineByPeriodicity")]
        public IHttpActionResult GetMagazineByPeriodicity(string periodicity)
        {
            var magazines = _periodicMagazineService.GetMagazineByPeriodicity(periodicity);
            return Ok(magazines);
        }

        [HttpGet]
        [Route("magazineSubscription")]
        public IHttpActionResult GetMagazineSubscriptionInYear()
        {
            var years = _periodicMagazineService.GetMagazineSubscriptionInYear();
            return Ok(years);
        }

        [HttpGet]
        [Route("magazineBySubscription/{year}")]
        public IHttpActionResult GetMagazineBySubscription(int year)
        {
            var magazines = _periodicMagazineService.GetMagazineBySubscription(year);
            return Ok(magazines);
        }

    }
}

File: LibraryBackEnd/app/components/Statistics/statisticsController.js
StatisticsModule.controller("StatisticsController", [
    "StatisticsFactory",
    "AppService",
    "$scope",
    "$location",
    "$route",
    "$routeParams",
    "$timeout",
    "$rootScope",
    "toastr",
    "$window",
    function (StatisticsFactory, AppService, $scope, $location, $route, $routeParams, $timeout, $rootScope,
        toastr, $window) {

        var me = this;

        $scope.oneAtATime = true;

        me.booksInCourse = [];

        me.booksInSubject = [];

        me.booksBaughtInYear = [];

        me.studentsInCourse = [];

        me.studentsInYear = [];

        me.bookTitles = [];

        me.bookTypes = [];

        me.magazineTiles = [];

        me.magazineCourses = [];

        me.magazineTypes = [];

        me.periodicity = [];

        me.subscription = [];

        me.returnedBooks = {
            year: [],
            books: []
        };

        me.returnedMagazines = {
            year: [],
            magazines: []
        };

        me.config = function () {
            me.dtOptions = AppService.dataTableWithOutFunction();
        }

        me.loadData = function () {
            me.loadBooksInCourse();
            me.loadBooksInSubject();
            me.loadBookBaughtInYear();
            me.loadStudentsInCourse();
            me.loadStudentsInYear();
            me.loadBooksTitles();
            me.loadBookTypes();
            me.loadMagazineTitleData();
            me.loadMagazineCourseData();
            me.loadMagazineTypeData();
            me.loadMagazinePeriodicityData();
            me.loadMagazineSubscriptionData();
            me.loadReturnedBooksInYear();
            me.loadReturnedMagazinesInYear();
        }

        me.loadBooksInCourse = function () {
            me.config();
            if (me.booksInCourse.length > 0)
                return;
            StatisticsFactory.getBooksInCourse().then(function (response) {
                me.booksInCourse = response;
            });
        }

        me.loadBooksInSubject = function () {
            me.config();
            if (me.booksInSubject.length > 0)
                return;
            StatisticsFactory.getBooksInSubject().then(function (response) {
                me.booksInSubject = response;
            });
        }

        me.loadBookBaughtInYear = function () {
            me.config();
            if (me.booksBaughtInYear.length > 0)
                return;
            StatisticsFactory.getBookBaughtInYear().then(function (response) {
                me.booksBaughtInYear = response;
            });
        }

        me.loadStudentsInCourse = function () {
            me.config();
            if (me.studentsInCourse.length > 0)
                return;
            StatisticsFactory.getStudentsInCourse().then(function (response) {
                me.studentsInCourse = response;

            });
        }

        me.loadStudentsInYear = function () {
            me.config();
            if (me.studentsInYear.length > 0)
                return;
            StatisticsFactory.getStudentsInYear().then(function (response) {
                me.studentsInYear = response;
            });
        }

        me.loadBooksTitles = function () {
            me.config();
            if (me.bookTitles.length > 0)
                return;
            StatisticsFactory.getBooksTitles().then(function (response) {
                me.bookTitles = response;
            });
        }

        me.loadBookTypes = function () {
            me.config();
            if (me.bookTypes.length > 0)
                return;
            StatisticsFactory.getBookTypes().then(function (response) {
                me.bookTypes = response;
            });
        }

        me.loadMagazineTitleData = function () {
            me.config();
            if (me.magazineTiles.length > 0)
                return;
            StatisticsFactory.getMagazineTitles().then(function (response) {
                me.magazineTitles = response;
            });
        }

        me.loadDataByCourse = function () {
            me.config();
            $rootScope.isBusy = true;
            me.courseName = $routeParams.courseName;
            StatisticsFactory.getBooksByCourse(me.courseName).then(function (response) {
                me.books = response;
                $rootScope.isBusy = false;
            }, function () {
                $rootScope.isBusy = false;
                toastr.error("Error fetching data.");
            });

        }

        me.loadDataBySubject = function () {
            me.config();
            $rootScope.isBusy = true;
            me.subject = $routeParams.subject;
            StatisticsFactory.getBooksBySubject(me.subject).then(function (response) {
                me.books = response;
                $rootScope.isBusy = false;
            }, function () {
                $rootScope.isBusy = false;
                toastr.error("Error fetching data.");
            });
        }

        me.loadDataByYear = function () {
            me.config();
            $rootScope.isBusy = true;
            me.year = $routeParams.year;
            StatisticsFactory.getBookBaughtByYear(me.year).then(function (response) {
                me.books = response;
                $rootScope.isBusy = false;
            }, function () {
                $rootScope.isBusy = false;
                toastr.error("Error fetching data.");
            });
        }

        me.loadDataByTitle = function () {
            me.config();
            $rootScope.isBusy = true;
            me.title = $routeParams.title;
            StatisticsFactory.getBooksByTitle(me.title).then(function (response) {
                me.books = response;
                $rootScope.isBusy = false;
            }, function () {
                $rootScope.isBusy = false;
                toastr.error("Error fetching data.");
            });
        }

        me.loadStudentDataByCourse = function () {
            me.config();
            $rootScope.isBusy = true;
            me.courseName = $routeParams.courseName;
            console.log(me.courseName);
            StatisticsFactory.getStudentsByCourse(me.courseName).then(function (response) {
                me.students = response;
                $rootScope.isBusy = false;
            }, function () {
                $rootScope.isBusy = false;
                toastr.error("Error fetching data.");
            });
        }

        me.loadStudentDataByYear = function () {
            me.config();
            $rootScope.isBusy = true;
            me.year = $routeParams.year;
            StatisticsFactory.getStudentsByYear(me.year).then(function (response) {
                me.students = response;
                $rootScope.isBusy = false;
            }, function () {
                $rootScope.isBusy = false;
                toastr.error("Error fetching data.");
            });
        }

        me.loadBookDataByTypes = function () {
            me.config();
            $rootScope.isBusy = true;
            me.type = $routeParams.type;
            StatisticsFactory.getBookByType(me.type).then(function (response) {
                me.books = response;
                $rootScope.isBusy = false;
            }, function () {
                $rootScope.isBusy = false;
                toastr.error("Error fetching data.");
            });
        }

        me.loadMagazineDataByTitle = function () {
            me.config();
            $rootScope.isBusy = true;
            me.title = $routeParams.title;
            StatisticsFactory.getMagazineByTitle(me.title).then(function (response) {
                me.magazines = response;
                $rootScope.isBusy = false;
            }, function () {
                $rootScope.isBusy = false;
                toastr.error("Error fetching data.");
            });
        }

        me.loadMagazineCourseData = function () {
            me.config();
            if (me.magazineCourses.length > 0)
                return;
            StatisticsFactory.getMagazineCourse().then(function (response) {
                me.magazineCourses = response;
            });
        }

        me.loadMagazineDataByCourse = function () {
            me.config();
            $rootScope.isBusy = true;
            me.course = $routeParams.course;
            StatisticsFactory.getMagazineByCourse(me.course).then(function (response) {
                me.magazines = response;
                $rootScope.isBusy = false;
            }, function () {
                $rootScope.isBusy = false;
                toastr.error("Error fetching data.");
            });
        }

        me.loadMagazineTypeData = function () {
            me.config();
            if (me.magazineTypes.length > 0)
                return;
            StatisticsFactory.getMagazineTypes().then(function (response) {
                me.magazineTypes = response;
            });
        }

        me.loadMagazineDataByType = function () {
            me.config();
            $rootScope.isBusy = true;
            me.type = $routeParams.type;
            StatisticsFactory.getMagazineByTypes(me.type).then(function (response) {
                me.magazines = response;
                $rootScope.isBusy = false;
            }, function () {
                $rootScope.isBusy = false;
                toastr.error("Error fetching data.");
            });
        }

        me.loadMagazinePeriodicityData = function () {
            me.config();
            if (me.periodicity.length > 0)
                return;
            StatisticsFactory.getMagazinePeriodicity().then(function (response) {
                me.periodicity = response;
            });
        }

        me.loadMagazineDataByPeriodicity = function () {
            me.config();
            $rootScope.isBusy = true;
            me.periodic = $routeParams.periodic;
            StatisticsFactory.getMagazineByPeriodicity(me.periodic).then(function (response) {
                me.magazines = response;
                $rootScope.isBusy = false;
            }, function () {
                $rootScope.isBusy = false;
                toastr.error("Error fetching data.");
            });
        }

        me.loadMagazineSubscriptionData = function () {
            me.config();
            if (me.subscription.length > 0)
                return;
            StatisticsFactory.getMagazineSubscriptionInYear().then(function (response) {
                me.subscription = response;
            });
        }

        me.loadMagazineDataBySubscription = function () {
            me.config();
            $rootScope.isBusy = true;
            me.year = $routeParams.year;
            StatisticsFactory.getMagazineBySubscription(me.year).then(function (response) {
                me.magazines = response;
                $rootScope.isBusy = false;
            }, function () {
                $rootScope.isBusy = false;
                toastr.error("Error fetching data.");
            });
        }

        me.loadReturnedBooksInYear = function () {
            me.config();
            $rootScope.isBusy = true;
            StatisticsFactory.getReturnedBooksInYear().then(function (response) {
                me.returnedBooks.year = response;
                $rootScope.isBusy = false;
            }, function () {
                $rootScope.isBusy = false;
                toastr.error("Error fetching data.");
            });
        }

        me.loadReturnedBooksByYear = function () {
            me.config();
            $rootScope.isBusy = true;
            me.year = $routeParams.year;
            StatisticsFactory.getReturnedBooksByYear(me.year).then(function (response) {
                me.returnedBooks.books = response;
                $rootScope.isBusy = false;
            }, function () {
                $rootScope.isBusy = false;
                toastr.error("Error fetching data.");
            });

        }


        me.loadReturnedMagazinesInYear = function () {
            me.config();
            $rootScope.isBusy = true;
            StatisticsFactory.getReturnedMagazinesInYear().then(function (response) {
                me.returnedMagazines.year = response;
                $rootScope.isBusy = false;
            }, function () {
                $rootScope.isBusy = false;
                toastr.error("Error fetching data.");
            });
        }

        me.loadReturnedMagazinesByYear = function () {
            me.config();
            $rootScope.isBusy = true;
            me.year = $routeParams.year;
            StatisticsFactory.getReturnedMagazinesByYear(me.year).then(function (response) {
                me.returnedMagazines.magazines = response;
                $rootScope.isBusy = false;
            }, function () {
                $rootScope.isBusy = false;
                toastr.error("Error fetching data.");
            });

        }

    }
]);